src/persist-txn/src/txn_read.rs

```diff
@@ -0,0 +1,280 @@
+        assert!(self.progress_exclusive > ts);
+        self.datas
+            .get(data_id)
+            .and_then(|all| {
+                let ret = all.iter().rev().find(|x| **x <= ts).copied();
+                debug!(
+                    "to_data_inclusive {:.9} t={} ret={:?}: all={:?}",
+                    data_id.to_string(),
+                    ts,
+                    ret,
+                    all,
+                );
+                ret
+            })
+            .ok_or(NotRegistered {
+                data_id: *data_id,
+                ts: self.progress_exclusive,
+            })
+    }
+
+    /// Returns the batches needing cleanup as of the current progress.
+    pub(crate) fn live_batches(&self) -> impl Iterator<Item = &(ShardId, String, u64)> {
+        self.batches.values()
+    }
+
+    /// Invariant: afterward, self.progress_exclusive will be > ts
+    pub(crate) async fn update_past(&mut self, ts: u64) {
+        self.update(|progress_exclusive| progress_exclusive > ts)
+            .await;
+        debug_assert!(self.progress_exclusive > ts);
+        debug_assert_eq!(self.validate(), Ok(()));
+    }
+
+    /// Invariant: afterward, self.progress_exclusive will be >= ts
+    pub(crate) async fn update_to(&mut self, ts: u64) {
+        self.update(|progress_exclusive| progress_exclusive >= ts)
+            .await;
+        debug_assert!(self.progress_exclusive >= ts);
+        debug_assert_eq!(self.validate(), Ok(()));
+    }
+
+    async fn update<F: Fn(u64) -> bool>(&mut self, done: F) {
+        while !done(self.progress_exclusive) {
+            let events = self.subscribe.fetch_next().await;
+            for event in events {
+                let mut updates = match event {
+                    ListenEvent::Progress(frontier) => {
+                        self.progress_exclusive = *frontier
+                            .as_option()
+                            .expect("nothing should close the txns shard");
+                        continue;
+                    }
+                    ListenEvent::Updates(updates) => updates,
+                };
+                // Persist emits the times sorted by little endian encoding,
+                // which is not what we want
+                updates.sort_by(|(akv, at, _), (bkv, bt, _)| (at, akv).cmp(&(bt, bkv)));
+                for ((k, v), t, d) in updates {
+                    let data_id = k.expect("valid data_id");
+                    let batch = v.expect("valid batch");
+                    self.push(data_id, batch, t, d);
+                }
+            }
+        }
+        debug!(
+            "cache correct before {} len={} least_ts={:?}",
+            self.progress_exclusive,
+            self.batches.len(),
+            self.batches.first_key_value().map(|(_, (_, _, ts))| ts),
+        );
+    }
+
+    fn push(&mut self, data_id: ShardId, batch: String, ts: u64, diff: i64) {
+        if batch.is_empty() {
+            assert_eq!(diff, 1);
```

Comment:
What about `-1`? Is it worth using a different type if we're only expecting two values?